src/Symfony/Component/HttpClient/RetryHttpClient.php

```diff
@@ -0,0 +1,120 @@
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+/**
+ * Automatically retries the failing requests.
+ *
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+class RetryHttpClient implements HttpClientInterface
+{
+    use AsyncDecoratorTrait;
+
+    private $decider;
+    private $strategy;
+    private $maxRetries;
+    private $logger;
+
+    /**
+     * @param int $maxRetries The maximum number of times to retry
+     */
+    public function __construct(HttpClientInterface $client, RetryDeciderInterface $decider = null, RetryBackOffInterface $strategy = null, int $maxRetries = 3, LoggerInterface $logger = null)
+    {
+        $this->client = $client;
+        $this->decider = $decider ?? new HttpCodeDecider();
+        $this->strategy = $strategy ?? new ExponentialBackOff();
+        $this->maxRetries = $maxRetries;
+        $this->logger = $logger ?: new NullLogger();
+    }
+
+    public function request(string $method, string $url, array $options = []): ResponseInterface
+    {
+        $retryCount = 0;
+
+        return new AsyncResponse($this->client, $method, $url, $options, function (ChunkInterface $chunk, AsyncContext $context) use ($method, $url, $options, &$retryCount) {
+            $exception = null;
+            try {
+                if ($chunk->isTimeout() || null !== $chunk->getInformationalStatus()) {
+                    yield $chunk;
+
+                    return;
+                }
+
+                // only retry first chunk
+                if (!$chunk->isFirst()) {
+                    $context->passthru();
+                    yield $chunk;
+
+                    return;
+                }
+            } catch (TransportExceptionInterface $exception) {
+                // catch TransportExceptionInterface to send it to strategy.
+            }
+
+            if ($retryCount >= $this->maxRetries || !$this->decider->shouldRetry($method, $url, $options, $partialResponse = new MockResponse($context->getContent(), ['http_code' => $statusCode = $context->getStatusCode(), 'headers' => $headers = $context->getHeaders()]), $exception)) {
+                $context->passthru();
+                yield $chunk;
+
+                return;
+            }
+
+            $context->setInfo('retry_count', $retryCount);
+            $context->getResponse()->cancel();
+
+            $delay = $this->getDelayFromHeader($headers) ?? $this->strategy->getDelay($retryCount, $method, $url, $options, $partialResponse, $exception);
+            ++$retryCount;
+
+            $this->logger->info('Error returned by the server. Retrying #{retryCount} using {delay} ms delay: '.($exception ? $exception->getMessage() : 'StatusCode: '.$statusCode), [
+                'retryCount' => $retryCount,
+                'delay' => $delay,
+            ]);
+
+            try {
+                $chunk->getContent();
+            } catch (TransportExceptionInterface $exception) {
+                // silent the failsafe exception
```

Comment:
are you sure this try/catch is needed? what happens if you remove it?

Reply:
indeed, it's not needed since I don't clone the chunk anymore. Removed